src/Common/src/TypeSystem/Common/MetadataType.cs

```diff
@@ -59,6 +59,26 @@ public override bool HasStaticConstructor
         /// Returns true if the type has given custom attribute.
         /// </summary>
         public abstract bool HasCustomAttribute(string attributeNamespace, string attributeName);
+
+        /// <summary>
+        /// Adapt current to the <paramref name="typeInstantiation"/> context. For example, if you have C <T> and a
+        /// context [int], it will yield C <int>.
+        /// </summary>
+        /// <param name="typeInstantiation">Context</param>
+        /// <param name="methodInstantiation">Not used for types.</param>
+        /// <returns></returns>
+        public override TypeDesc InstantiateSignature(Instantiation typeInstantiation, Instantiation methodInstantiation)
+        {
+            // Type is either not generic, or no valid context is provided.
+            if (!HasInstantiation || typeInstantiation.IsNull)
+            {
+                return this;
+            }
+            else
+            {
+                return new InstantiatedType(this, typeInstantiation);
```

Comment:
BTW: The typesystem objects should be always created through factory so that there is always only one for each identity ever created.